File: security/preflight.py
#!/usr/bin/env python3
"""
Preflight Security Check — ARGOS unified secret validation.

Lee `required_secrets.json` y valida que las variables de entorno
requeridas estén presentes y no vacías según el modo de operación.

USO:
    python3 security/preflight.py                    # check all services
    python3 security/preflight.py data-engine        # check only data-engine
    python3 security/preflight.py analytics-engine   # check only analytics-engine
    ENVIRONMENT_MODE=LIVE python3 security/preflight.py

EXIT CODES:
    0  → todos los checks pasaron
    1  → falta al menos un secret requerido

Arquitectura:
    - Infisical (DEV/STAGING) / systemd (PROD) inyectan las vars.
    - Este script verifica que estén presentes antes de iniciar.
    - No depende de dotenv. Lee directo de os.environ.
"""
from __future__ import annotations

import json
import os
import sys
from pathlib import Path


def load_schema() -> dict:
    schema_path = Path(__file__).resolve().parent / "required_secrets.json"
    with open(schema_path) as f:
        return json.load(f)


def get_required_vars(schema: dict, mode: str) -> list[str]:
    """Return the list of required env var names for the given mode."""
    mode_cfg = schema.get("mode_requirements", {}).get(mode, {})
    required_groups = mode_cfg.get("required_groups", [])

    required: list[str] = []
    for group_name in required_groups:
        group = schema.get(group_name, [])
        required.extend(group)
    return required


def check_vars(required: list[str]) -> list[str]:
    """Return list of missing/empty var names."""
    errors: list[str] = []
    for var in required:
        val = os.environ.get(var)
        if not val or val.strip() == "":
            errors.append(var)
    return errors


def preflight_check(service: str | None = None) -> list[str]:
    """Run preflight for given service (or all). Returns list of errors."""
    schema = load_schema()
    mode = os.environ.get("ENVIRONMENT_MODE", "BACKTESTING").upper()

    required = get_required_vars(schema, mode)
    return check_vars(required)


def main() -> None:
    service = sys.argv[1] if len(sys.argv) > 1 else "all"
    mode = os.environ.get("ENVIRONMENT_MODE", "BACKTESTING").upper()

    errors = preflight_check(service)

    if errors:
        print(
            f"SECURITY PREFLIGHT FAILED [{service}] mode={mode}",
            file=sys.stderr,
        )
        print(f"  Missing required secrets: {', '.join(errors)}", file=sys.stderr)
        print(
            "  Injection: use 'infisical run' (DEV/STAGING) or set env vars (PROD)",
            file=sys.stderr,
        )
        sys.exit(1)

    print(
        f"SECURITY PREFLIGHT PASSED [{service}] mode={mode}",
        file=sys.stderr,
    )
    sys.exit(0)


if __name__ == "__main__":
    main()

File: apps/data-engine/src/infrastructure/security/preflight.ts
/**
 * Preflight Security Check — ARGOS data-engine (TypeScript).
 *
 * Lee `required_secrets.json` (schema compartido con Python) y valida
 * que las variables de entorno requeridas estén presentes según el modo.
 *
 * USO:
 *   import { preflightCheck } from "./infrastructure/security/preflight"
 *   preflightCheck()  // llama antes de NestFactory.create()
 *
 * Arquitectura:
 *   - Infisical (DEV/STAGING) / systemd (PROD) inyectan las vars.
 *   - Este módulo verifica que estén presentes antes de iniciar.
 *   - No depende de dotenv. Lee directo de process.env.
 */

import { readFileSync, existsSync } from "fs"
import { join, resolve } from "path"

interface SecretsSchema {
  production: string[]
  testnet: string[]
  notifications: string[]
  optional: string[]
  mode_requirements: Record<
    string,
    { required_groups: string[]; description: string }
  >
}

function loadSchema(): SecretsSchema {
  const searchPaths = [
    join(__dirname, "..", "..", "..", "..", "..", "security", "required_secrets.json"),
    join(process.cwd(), "security", "required_secrets.json"),
    resolve("/etc/argos/required_secrets.json"),
  ]

  for (const p of searchPaths) {
    if (existsSync(p)) {
      try {
        const raw = readFileSync(p, "utf8")
        return JSON.parse(raw) as SecretsSchema
      } catch {
        // try next path
      }
    }
  }

  // Fallback: schema embebido (último recurso)
  return {
    production: ["EXCHANGE_API_KEY", "EXCHANGE_API_SECRET", "ARGOS_BROKER_URL"],
    testnet: ["BINANCE_TESTNET_API_KEY", "BINANCE_TESTNET_SECRET"],
    notifications: ["TELEGRAM_BOT_TOKEN", "TELEGRAM_CHAT_ID"],
    optional: [
      "EXCHANGE_PASSPHRASE",
      "EXCHANGE_ID",
      "EXCHANGE_WS_URL",
      "DISCORD_WEBHOOK_URL",
      "USE_PYTORCH",
      "RISK_PCT",
      "DRAWDOWN_PCT",
      "ARGOS_HISTORICAL_DIR",
    ],
    mode_requirements: {
      BACKTESTING: { required_groups: [], description: "No secrets required" },
      PAPER_TRADING: {
        required_groups: ["testnet", "notifications"],
        description: "Requires testnet + notifications",
      },
      LIVE: {
        required_groups: ["production", "notifications"],
        description: "Requires all secrets",
      },
      LIVE_SIMULATION: {
        required_groups: ["testnet", "notifications"],
        description: "Simulation mode — requires testnet + notifications",
      },
    },
  }
}

function getRequiredVars(schema: SecretsSchema, mode: string): string[] {
  const modeCfg = schema.mode_requirements[mode]
  if (!modeCfg) return []

  const required: string[] = []
  for (const groupName of modeCfg.required_groups) {
    const group = (schema as unknown as Record<string, string[]>)[groupName]
    if (group) {
      required.push(...group)
    }
  }
  return required
}

function checkVars(required: string[]): string[] {
  return required.filter((v) => {
    const val = process.env[v]
    return !val || val.trim() === ""
  })
}

export function preflightCheck(): void {
  const mode = (process.env.ENVIRONMENT_MODE ?? "BACKTESTING").toUpperCase()
  const schema = loadSchema()
  const required = getRequiredVars(schema, mode)
  const missing = checkVars(required)

  if (missing.length > 0) {
    console.error(
      `[security] PREFLIGHT FAILED [data-engine] mode=${mode}`,
    )
    console.error(`  Missing required secrets: ${missing.join(", ")}`)
    console.error(
      "  Injection: use 'infisical run' (DEV/STAGING) or set env vars (PROD)",
    )
    process.exit(1)
  }

  // eslint-disable-next-line no-console
  console.info(`[security] PREFLIGHT PASSED [data-engine] mode=${mode}`)
}
